model-server/src/main/kotlin/org/modelix/model/server/handlers/ModelReplicationServer.kt

```diff
@@ -68,171 +70,222 @@ class ModelReplicationServer(val repositoriesManager: RepositoriesManager) {
+            call.respondText(newVersionHash)
+        }
+
+        // TODO: migrate to use openapi or corresponding approach
+        route("v2/repositories/{repository}/branches/{branch}") {
+            fun ApplicationCall.repositoryId() = RepositoryId(parameters["repository"]!!)
+            fun PipelineContext<Unit, ApplicationCall>.repositoryId() = call.repositoryId()
+            fun ApplicationCall.branchRef() = repositoryId().getBranchReference(parameters["branch"]!!)
+            fun PipelineContext<Unit, ApplicationCall>.branchRef() = call.branchRef()
+
+            webSocket("listen") {
+                var lastVersionHash = call.request.queryParameters["lastKnown"]
+                while (coroutineContext[Job]?.isCancelled == false) {
+                    val newVersionHash =
+                        repositoriesManager.pollVersionHash(call.branchRef(), lastVersionHash)
+                    val delta = VersionDelta(
+                        newVersionHash,
+                        lastVersionHash,
+                        objectsMap = repositoriesManager.computeDelta(newVersionHash, lastVersionHash),
                     )
-                    return@get
+                    delta.checkObjectHashes()
+                    send(Json.encodeToString(delta))
+                    lastVersionHash = newVersionHash
                 }
-                call.respondDelta(versionHash, baseVersionHash)
             }
-            get("{versionHash}/history/{oldestVersionHash}") {
-                TODO()
+        }
+
+        get<Paths.v2RepositoriesRepositoryVersionsVersionHashGet> {
+            fun ApplicationCall.repositoryId() = RepositoryId(parameters["repository"]!!)
+            fun PipelineContext<Unit, ApplicationCall>.repositoryId() = call.repositoryId()
+
+            // TODO permission check on the repository ID is not sufficient, because the client could
+            //      provide any repository ID to access a version inside a different repository.
+            //      A check if the version belongs to the repository is required.
+            val baseVersionHash = call.request.queryParameters["lastKnown"]
+            val versionHash = call.parameters["versionHash"]!!
+            if (storeClient[versionHash] == null) {
+                call.respondText(
+                    "Version '$versionHash' doesn't exist",
+                    status = HttpStatusCode.NotFound,
+                )
+                return@get
             }
+            call.respondDelta(versionHash, baseVersionHash)
+        }
+        get<Paths.v2RepositoriesRepositoryVersionsVersionHashHistoryOldestVersionHashGet> {
+            fun ApplicationCall.repositoryId() = RepositoryId(parameters["repository"]!!)
+            fun PipelineContext<Unit, ApplicationCall>.repositoryId() = call.repositoryId()
+
+            TODO()
+        }
+
+        post<Paths.v2RepositoriesRepositoryBranchesBranchQueryPost> {
```

Comment:
Can these ugly names be made more pretty by setting an `operationId` (cf. https://swagger.io/docs/specification/paths-and-operations/) in the spec?

Reply:
Generally, yes. But practically I do not even know the purpose of all available endpoints. This will be a dedicated next step. I will add those I know and can semantically understand. The others will need refinement later on (just like the entire API...)

Reply:
Can't we just use this PR to sort out good names?